### ISSUE TITLE
OIDC state cookie re-validated on any endpoint

### ISSUE THREAD

**Issue by DCCSKrezovic** [quarkusio/quarkus#29162]
### Describe the bug

If my redirect request fails for whatever reason, q_state cookie will remain. On any next request to any endpoint, it will try to redo the code flow... This is, of course, impossible, as there's no state parameter present and will always return an error.

Method

```
public Uni<SecurityIdentity> authenticate(RoutingContext context,
            IdentityProviderManager identityProviderManager, OidcTenantConfig oidcTenantConfig)
```

`CodeAuthenticationMechanism` says:

```
        // if the state cookie is available then try to complete the code flow and start a new session
        if (stateCookie != null) {
            LOG.debug("State cookie is present, processing an expected redirect from the OIDC provider");
```

However, this does not make sense on any endpoint than the configured redirect url as defined by

`quarkus.oidc.authentication.redirect-path`

At the moment, it tries to re-do the authentication on any endpoint, which is wrong. A call to index.html or any other endpoint that is not the redirect path will throw an error in the console

```
2022-11-09 22:52:13,663 ERROR [io.qua.oid.run.CodeAuthenticationMechanism] (vert.x-eventloop-thread-3) State verification has failed, completing the code flow with HTTP status 401
```

### Expected behavior

Cookie is only used on the redirect URI path defined by configuration.

### Actual behavior

Cookie is read and disaster occurs on any path

### Quarkus version or git rev

2.13.4.Final

**Comment by DCCSKrezovic** [comment]
@sberyozkin  this is actually a variant of the bug it was duplicated by. In my case, it happens with two "nearly concurrent" login requests, one triggered by fetch(), another by browser.

fetch() tries to access protected backend resource, redirect is triggered and state cookie is created. Redirect fails due to CORS error
Browser tries to access a protected URL using state cookie created by fetch() but the endpoint is not redirect url ... 401.

I would like to address this issue specifically by only triggering validation of the state cookie on redirect url. Would that be fine for you?

**Comment by DCCSKrezovic** [comment]
From what I see, there is already such check for logout path

```
    private boolean isLogout(RoutingContext context, TenantConfigContext configContext) {
        Optional<String> logoutPath = configContext.oidcConfig.logout.path;

        if (logoutPath.isPresent()) {
            return context.request().absoluteURI().equals(
                    buildUri(context, false, logoutPath.get()));
        }

        return false;
    }
```

There should be an equivalent for login url as well and that should be checked before examining the state cookie. Again, examining the state cookie and handling the code request only makes sense on the redirect url.

**Comment by sberyozkin** [comment]
@DCCSKrezovic Logout case is totally different, this request it is coming from the user, not after a redirect from somewhere

**Comment by DCCSKrezovic** [comment]
I would like to fix this specific bug myself ASAP as it's quite urgent. Can you point me in some directions you'd like to go with the mentioned task? I was thinking of:

- Checking state cookie only on login endpoint
- Returning 401 if no cookie is found on login endpoint

This seems sufficient and it's according to any other OIDC implementation I saw so far.

Also, the state cookie SHOULD be removed after successful authentication anyway ... but isn't that already the case?

**Comment by sberyozkin** [comment]
@DCCSKrezovic I've already said it before but I do not care how some other implementations deal with it - Quarkus OIDC approach is conservative and supporting concurrent/parallel logins on behalf of the same user does not look like a mainstream or otherwise critical case. 
I've already started working on it and have a branch in progress but other higher priorities issues such as those affecting the typical flows have had to be dealt with. The work I've started doing on this branch is sensitive, and I don't want to mess up and fix it in a rush. FYI, I'm against removing a state check on anything but the redirect URI, it will remain but users will be able to restrict to the redirect-uri only.

You are welcome to work on other issues
Cheers
